core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -633,6 +633,7 @@ private boolean expandStar(List<SqlNode> selectItems, Set<String> aliases,
     if (!identifier.isStar()) {
       return false;
     }
+    final int offset = selectItems.size();
```

Comment:
If `offset` is judged here, then all fields before * will not permutation.
I recommend improving the `calculatePermuteOffset` method. like
```
int originSize = selectItems.size();
......
int offset = Math.min(calculatePermuteOffset(selectItems), originSize);
```

Reply:
Updated.

Reply:
@YiwenWu Can you take another look?